File: src/216.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> res;
        vector<int> candidates = {1,2,3,4,5,6,7,8,9};
        for (int i = 0; i < candidates.size(); i ++) {
            printf("%d,", candidates[i]);
        }
        
        vector<int> temp;
        dfs(candidates, n, temp, res, 0, k);
         
        return res;
    }
    
    void dfs(vector<int> &candidates, int target, vector<int> &temp, vector<vector<int>> &res, int index, int k) {
        if (temp.size() > k || target < 0) {
            return;
        }
        if (temp.size() == k && target == 0) {
            res.push_back(temp);
        } else {
            for (int i = index; i < candidates.size(); i ++) {
                temp.push_back(candidates[i]);
                dfs(candidates, target - candidates[i], temp, res, i + 1, k);
                temp.pop_back();
            }
        }
        
    }
};

File: src/216.py
class Solution(object):
    def combinationSum3(self, k, n):
        """
        :type k: int
        :type n: int
        :rtype: List[List[int]]
        """
        res = []
        candidates = [1,2,3,4,5,6,7,8,9]
        self.dfs([], res, 0, k, n, candidates)
        return res
    
    
    def dfs(self, temp, res, startIndex, k, target, candidates):
        if len(temp) > k or target < 0:
            return
        if len(temp) == k and target == 0:
            res.append([] + temp)
        else:
            for i in range(startIndex, len(candidates)):
                temp.append(candidates[i])
                # print(temp)
                self.dfs(temp, res, i + 1, k, target - candidates[i], candidates)
                temp.pop()